internal/impl/devbox.go

```diff
@@ -981,3 +1028,14 @@ func (d *Devbox) NixBins(ctx context.Context) ([]string, error) {
 	}
 	return lo.Values(bins), nil
 }
+
+func addHashToEnv(env map[string]string) map[string]string {
+	data := []byte{}
+	for k, v := range env {
+		data = append(data, []byte(k)...)
+		data = append(data, []byte(v)...)
+	}
+	hash := md5.Sum(data)
+	env[devboxShellEnvHashVarName] = hex.EncodeToString(hash[:])
+	return env
+}
```

Comment:
`data` can be removed by hashing as you go:

```suggestion
func addHashToEnv(env map[string]string) map[string]string {
	h := md5.New()
	for k, v := range env {
		h.Write([]byte(k))
		h.Write([]byte(v))
	}
	env[devboxShellEnvHashVarName] = hex.EncodeToString(md5.Sum(data))
	return env
}
```

If you want, the return value can also be removed since the function is modifying the map directly.